# Comparing Left, Right, and Midpoint Riemann Sums to the Exact Value of a Function's Area Under the Curve

Below is the comparison of the exact value of a function's area under the curve and the riemann sum values using different amounts of rectangles. Functions for left, right, and midpoint riemann sums were written so that any value of a, c, and n can be entered into the function. Rectangle values of 10,15, and 100 were used for each.

## a=3, c=4, n=10,15,100

## Integrate the function f(x)=$a^2$$xe^{-ax}$ over the invterval 0$\le$x$\le$c; for all values a>0

$\int\limits_0^4$$a^2$$xe^{-ax}dx$ =
$\int\limits_0^4$$9$$xe^{-3x}dx$ =
$9\int\limits_0^4$$xe^{-3x}dx$

$u=x$    $dv=e^-$$^3$$^x$

$du=dx$    $v=$ $\frac{-e^{-3x}}{3}$

$9$[$\frac{-xe^{-3x}}{3}$-$\int\limits_0^4$$\frac{-e^{-3x}}{3}dx$]

$9$[$\frac{-xe^{-3x}}{3}$+$^1/_3$$\int\limits_0^4$$e^{-3x}dx$]

$u= -3x$

$du= -3dx$

$9$[$\frac{-xe^{-3x}}{3}$-$^1/_9$$\int\limits_0^4$$e^udu$]

$9$[$\frac{-xe^{-3x}}{3}$-$\frac{e^{-3x}}{9}$] 0$\le$x$\le$4

$-3xe^{-3x}-e^{-3x}$ 0$\le$x$\le$4

[$-3(4)e^{-3(4)}-e^{-3(4)}$]$-$[$-3(0)e^{-3(0)}-e^{-3(0)}$]

The exact answer is $0.99992012$

## For the riemann sums, we rounded e to 2.718.

## Left Point Riemann Sum

In [2]:
def left_point(startinter,endinter,a,n):
    total=0
    for i in range(n):
        width=((endinter-startinter)/n)
        height=(a**2)*(((endinter-startinter)/n)*i)*(2.718**(-a*((endinter-startinter)/n)*i))
        area=height*width
        total+=area
    return total

print(round(left_point(0,4,3,10),4))
%timeit left_point(0,4,3,10)

0.8882
4.87 µs ± 19.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [3]:
print(round(left_point(0,4,3,15),4))
%timeit left_point(0,4,3,15)

0.9484
7.19 µs ± 24 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [4]:
print(round(left_point(0,4,3,100),4))
%timeit left_point(0,4,3,100)

0.9989
46.9 µs ± 178 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Right Point Riemann Sum

In [5]:
def right_point(startinter,endinter,a,n):
    total=0
    for i in range(n):
        width=((endinter-startinter)/n)
        height=(a**2)*(((endinter-startinter)/n)*(i+1))*(2.718**(-a*((endinter-startinter)/n)*(i+1)))
        area=height*width
        total+=area
    return total

print(round(right_point(0,4,3,10),4))
%timeit right_point(0,4,3,10)

0.8883
5.1 µs ± 46.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [6]:
print(round(right_point(0,4,3,15),4))
%timeit right_point(0,4,3,15)

0.9485
7.56 µs ± 12.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [7]:
print(round(right_point(0,4,3,100),4))
%timeit right_point(0,4,3,100)

0.9989
48.9 µs ± 368 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Midpoint Riemann Sum

In [8]:
def midpoint(startinter,endinter,a,n):
    total=0
    for i in range(n):
        width=(((endinter-startinter))/n)
        height=(a**2)*(((endinter-startinter)/n)*(i+.5))*(2.718**(-a*((endinter-startinter)/n)*(i+.5)))
        area=height*width
        total+=area
    return total

print(round(midpoint(0,4,3,10),4))
%timeit midpoint(0,4,3,10)

1.053
5.25 µs ± 21.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [9]:
print(round(midpoint(0,4,3,15),4))
%timeit midpoint(0,4,3,15)

1.0253
7.72 µs ± 9.47 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [10]:
print(round(midpoint(0,4,3,100),4))
%timeit midpoint(0,4,3,100)

1.0007
50.1 µs ± 44.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Final Results

Exact value: 0.99992012

| n=10 | n=15 | n=100 |
| :- | -: | :-: |
|left: 0.8882| left: 0.9484| left: 0.9989|
|right: 0.8883| right:  0.9485| right:0.9989|
|mid: 1.053| mid: 1.0253| mid: 1.0007|

Based on the table, right point sums produce the greatest output, left point sums are second, and midpoint sums produce the lowest output. In comparison to the exact value, using 100 rectangles and right point riemann sum gave us the closest answer. Using midpoint riemann sum gave an extreme underestimate. The more rectangles you use, the closer your answer is to the exact answer.

| Left Point Time | Right Point Time | Midpoint Time |
| :- | -: | :-: |
|n=10: 4.87 µs ± 19.8 ns per loop|n=10: 5.1 µs ± 46.2 ns per loop|n=10: 5.25 µs ± 21.6 ns per loop|
|n=15: 7.19 µs ± 24 ns per loop|n=15: 7.56 µs ± 12.4 ns per loop|n=15: 7.72 µs ± 9.47 ns per loop|
|n=100: 46.9 µs ± 178 ns per loop|n=100: 48.9 µs ± 368 ns per loop|n=100: 50.1 µs ± 44.4 ns per loop|

Using "timeit" us means microseconds. The fastest function runs for left point, then right point, then midpoint. Looking further, only using 10 rectangles runs the fastest and 100 runs the slowest. This makes sense considering the function runs through every value in the range. So, it is going to be faster when going through a range of 10 versus a range of 100

In conclusion, using a higher amount of rectangles and right riemann sum gives you the closest answer to the exact answer.